metabase/metabase
# Fix Add Series modal performance issues by moving the logic to the BE

Closes https://github.com/metabase/metabase/issues/28045

## Description

Fixes the add series modal that allows combining cards on dashboards. The logic that determines if two cards are compatible was implemented on the FE, it needed both cards and their metadata. So previously the add series modal loaded all the cards from the instance which made it unusable on larger instances. It initially hid structured incompatible questions, but for native queries, users had to try to select them to understand if they are compatible which was not great.

In this PR we added a new BE endpoint that exposes only compatible cards with a cursor-based pagination to avoid performance issues.

## How to verify

- Create a few compatible structured/native cards that have numeric/date dimensions for the x-axis
- Ensure you can combine them on a dashboard
- Ensure the results are paginated (to simplify testing this you can change the page size in `AddSeriesModal/QuestionList.tsx`

## API

Adding a new endpoint `GET /api/card/:id/series` with arguments:
- `last_cursor` : id of the last card from the last page
- `limit`: page_size
- `query`: filter cards by name
- `exclude_ids`: a list of card ids to exclude 
The endpoint will return only compatible cards to be used as a series for the given card.


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30321)
